### ISSUE TITLE
Missing Python dependency in 0.3.7 breaks new installs

### ISSUE THREAD

**Issue by garrettr** [freedomofpress/securedrop#1344]
https://github.com/freedomofpress/securedrop/pull/1306 updated the SecureDrop web application's Python dependencies to their latest versions, but in doing so also removed `pycrypto`, which was a mistake.

Without this dependency, SecureDrop cannot function, and a new install will fail with the following error:

```
TASK: [app | initialize sqlite db] ******************************************** 
failed: [app] => {"changed": true, "cmd": "su -s /bin/bash -c \"PYTHONPATH=/var/www/securedrop python -c 'import db; db.init_db()'\" www-data", "delta": "0:00:00.147033", "end": "[redacted timestamp]", "rc": 1, "start": "[redacted timestamp]"}
stderr: Traceback (most recent call last):
  File "<string>", line 1, in <module>
  File "/var/www/securedrop/db.py", line 26, in <module>
    import crypto_util
  File "/var/www/securedrop/crypto_util.py", line 6, in <module>
    from Crypto.Random import random
ImportError: No module named Crypto.Random

FATAL: all hosts have already failed -- aborting
```

This should have been caught in the pre-release testing for 0.3.7, but it was not. We investigated and the reason why it was not caught is:

1. We used our Vagrant-based virtual environments for all "new install of 0.3.7-rc1" testing
2. We used the official [ubuntu/trusty64](https://atlas.hashicorp.com/ubuntu/boxes/trusty64/) Vagrant images.
3. Apparently, these images include additional packages which are _not_ included in the Ubuntu 14.04 .iso. These packages include `python-crypto`, which "fixed" the missing dependency in `securedrop-app-code.deb`.

Short-term, we need to do a new release **immediately** to correct this issue, because it breaks all new installs of SecureDrop.

**Comment by garrettr** [comment]
@fowlslegs Please also verify that there are no further regressions in the Python dependencies due to #1306.

**Comment by garrettr** [comment]
```
vagrant@development:~$ aptitude why python-crypto
i   python-apport       Depends    python-launchpadlib (>= 1.5.7)
i A python-launchpadlib Depends    python-keyring (>= 0.5)
i A python-keyring      Recommends python-crypto
```

**Comment by fowlslegs** [comment]
What I did:

1. Rename release/0.3.8 -> release/0.3.9
2. Fork new release/0.3.8 off release/0.3.7.post1
3. Re-include `pycrypto` in new commit to release/0.3.8
4. Verified that no other regressions in Python dependencies have occured due to #1306. Specifically, I confirmed line for line that every other package had made it into the new requirements file in the diff for #1306.
5. Verified using the "base:release/0.3.8" search query that there are no open PRs against 0.3.8.
6. Noticed Github's automagic created milestone 0.3.9 and moved all the milestone markers from 0.3.8 to 0.3.9 for us.
7. Cherry-picked 28a4dc2b38db6f626669613935656c33daa20895 into release/0.3.9 to preserve commit history.

release/0.3.8 looks ready to me--we probably want to expedite the release process because it's a one-line change. Let me know if there's anything else you think needs to be done before we can move forward with this.

**Comment by garrettr** [comment]
> Noticed Github's automagic created milestone 0.3.9 and moved all the milestone markers from 0.3.8 to 0.3.9 for us.

That wasn't automagic, that was me 😆

**Comment by garrettr** [comment]
> 7. Cherry-picked 28a4dc2 into release/0.3.9 to preserve commit history.

You did this too soon. release/0.3.9 should continue linearly from 0.3.8, but you can't do that yet because we haven't release 0.3.8. At a minimum, you still need the commit that bumps the version numbers.

A better approach is to wait until 0.3.8 is released, than either rebase or merge it in release/0.3.9. Rebase is preferred to keep a linear history, although anybody who is working on release/0.3.9 may have to do some work to deal with the rebase (but it should be pretty minor since there aren't many commits on release/0.3.9 yet).

You should remove that cherry-picked commit.

**Comment by fowlslegs** [comment]
Good point. I just removed the cherry-picked commit.

> That wasn't automagic, that was me 😆 

You should have just let me believe. I wanted to believe.

**Comment by conorsch** [comment]
> automagic

Don't worry, I routinely mistake @garrettr for a commit-slinging, milestone-piledriving robot.

As for the _why_ of this regression, it's good that we've identified divergence in the base Vagrant image as the culprit. Testing a fresh install, rather than only upgrades, on hardware would have caught this.

In terms of virtual environments, though, we may be able to switch base boxes. Using the Ubuntu 14.04 provisionerless Bento Box from Chef ([full list of "bento" boxes](https://github.com/chef/bento)) causes the playbook to fail:

> TASK [app : Initialize sqlite database.] ***************************************
fatal: [app-prod]: FAILED! => {"changed": true, "cmd": "su -s /bin/bash -c \"PYTHONPATH=/var/www/securedrop python -c 'import db; db.init_db()'\" www-data", "delta": "0:00:00.408985", "end": "2016-06-22 23:46:57.626255", "failed": true, "rc": 1, "start": "2016-06-22 23:46:57.217270", "stderr": "Traceback (most recent call last):\n  File \"<string>\", line 1, in <module>\n  File \"/var/www/securedrop/db.py\", line 26, in <module>\n    import crypto_util\n  File \"/var/www/securedrop/crypto_util.py\", line 6, in <module>\n    from Crypto.Random import random\nImportError: No module named Crypto.Random", "stdout": "", "stdout_lines": [], "warnings": ["Consider using 'become', 'become_method', and 'become_user' rather than running su"]}
        to retry, use: --limit @install_files/ansible-base/securedrop-prod.retry

whereas testing with the trusty64 image does not.

**Comment by fowlslegs** [comment]
> the Ubuntu 14.04 provisionerless Bento Box from Chef 

Cool, so I just peeked at https://github.com/chef/bento/blob/master/ubuntu-14.04-amd64.json and that seems like exactly what we want to use in terms of reproducibility of test environment. OTOH, we recently switched to the 'ubuntu/trusty64' boxes because they are updated the most regularly. Bento Boxes at least do receive regular, but not as regular of updates--compare https://atlas.hashicorp.com/ubuntu/boxes/trusty64 to https://atlas.hashicorp.com/bento/boxes/ubuntu-14.04. I know you've proposed some ways @conorsch that we could avoid this, but I don't know if they would be worth the effort given the upcoming re-architecture process.

Also, the solution to a different issue described in https://github.com/freedomofpress/securedrop/issues/856#issuecomment-221106486 would have prevented this from slipping through as well.

**Comment by garrettr** [comment]
We verified that using the Ubuntu 14.04 bento box allows us to reproduce the install issue in the virtual environment, so it should be sufficient to allow us to test the fix as well. We included those changes to the Vagrantfile in [0.3.8-rc1](https://github.com/freedomofpress/securedrop/commit/585b103635de4a787add814997d6fcc5565cc55e) along with the fix to the Python dependencies, and are testing it now.

**Comment by garrettr** [comment]
This issue was fixed and QA'ed on release/0.3.8 and released as 0.3.8 yesterday. Official announcement: https://securedrop.org/news/securedrop-038-released

1. https://github.com/freedomofpress/securedrop/commit/28a4dc2b38db6f626669613935656c33daa20895 re-includes the missing pycrypto dependency.
2. https://github.com/freedomofpress/securedrop/commit/5497e883142476df999cb60345ab7d660275739e switches to using Vagrant bases boxes that are identical to an install from an Ubuntu .iso, which would have prevented this bug from escaping our release QA process (as explained in the description of this issue).
